tools/spin-ec2/aws.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"golang.org/x/mod/semver"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var versionRegex = regexp.MustCompile(`v[0-9]+\.[0-9]+\.[0-9]+`)
+
+// getAWSInstances return all the canary instance names mapped on ids.
+func getAWSInstances() (map[string]string, error) {
+	cmd := exec.Command("aws", []string{
+		"ec2",
+		"describe-instances",
+		"--output", "text",
+		"--filters", `Name=tag:Name,Values="` + hostPrefix + `:*"`, `Name=instance-state-name,Values=running`,
+		"--query", "Reservations[*].Instances[*].[Tags[?Key==`Name`], InstanceId]",
+	}...)
+
+	var out bytes.Buffer
+	var outErr bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Stderr = &outErr
+
+	err := cmd.Run()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get aws instances, error: %s", outErr.String())
+	}
+
+	return parseAWSInstances(out.String())
+}
+
+// parseAWSInstances parses the ec2 describe-instances response.
+func parseAWSInstances(response string) (map[string]string, error) {
+	result := make(map[string]string)
+	split := strings.Split(response, "\n")
+
+	for i := 0; i < len(split)-1; i += 2 {
+		instanceID := strings.TrimSpace(split[i])
+
+		if !strings.HasPrefix(split[i+1], "Name") {
+			return nil, fmt.Errorf("unexpected line received for instanceName: %s", split[i+1])
+		}
+		instanceName := strings.TrimSpace(strings.TrimPrefix(split[i+1], "Name"))
+
+		result[instanceID] = instanceName
+	}
+
+	return result, nil
+}
+
+// pruneInstances will terminate all instances except the ones that have
+// latest 2 infra-agent version installed.
+func pruneInstances(instances map[string]string, dryRun bool) error {
+
+	idsToTerminate, err := getInstancesToPrune(instances)
```

Comment:
Probably this should be called before we pass instances to pruneInstances? Like in main function.